### ISSUE TITLE
[Bug] Grad Setting Malfunctioning

### ISSUE THREAD

**Issue by AdamBelfki3** [ndif-team/nnsight#141]
I have come across an important issue related to setting `.grad` to a new tensor object. After setting `.grad` and saving it; the values returned by the Proxy reflect the original content of `.grad` before setting it to a new tensor, and is thus incorrect. This indicates that the update was somehow not set properly. Below is a minimal reproducible example, as well as some additional testing I conducted to help determine the exact provenance of the bug.

**Note:** It seems that the problem only directly affects the user’s ability to save and access the correct Grad Proxy value and does not affect the correctness of the gradient values in the interleaved execution of the model; the updated gradient values still propagate properly within the model and work as intended. 

Code:
``` Python3
with model.trace(input):
        model.layer1.output.requires_grad = True

        l1_grad_before = model.layer1.output.grad.clone().save()

        model.layer1.output.grad = model.layer1.output.grad.clone() * 2

        l1_grad_after = model.layer1.output.grad.save()

        loss = model.output.sum()
        loss.backward()

    print("L1_Grad_Before: ", l1_grad_before)
    print("L1_Grad_After: ", l1_grad_after)
```

Out:
```
L1_Grad_Before:  tensor([[-0.0336, -0.0316,  0.5292, -0.1040, -0.0569,  0.0234, -0.1658, -0.3092,
          0.0051,  0.3399]])
L1_Grad_After:  tensor([[-0.0336, -0.0316,  0.5292, -0.1040, -0.0569,  0.0234, -0.1658, -0.3092,
          0.0051,  0.3399]])
```
Intervention Graph:
![Layer 1 Grad Setting](https://github.com/ndif-team/nnsight/assets/57205266/54a3b07a-1511-41cb-944e-9f7fe08b4d8c)

This example shows that the Layer 1 Grad Proxies are identical and show the same values even after the modification. (`L1_Grad_After` should be double the value `L1_Grad_Before`)

---

To support the claim that the issue causing this bug is coming from the setting operation on the `.grad` attribute, I want to share with you the following two examples:

The **first example** demonstrates that even after `.grad` is set to a new tensor object, not only both saved proxies have the same values but also point to the same memory address. This indicates that the execution of the setting operation did not replace the reference of the old grad tensor with the new one.

Code:
``` Python3

with model.trace(input):
        model.layer1.output.requires_grad = True

        l1_grad_before = model.layer1.output.grad.save()

        model.layer1.output.grad = model.layer1.output.grad.clone() * 2

        l1_grad_after = model.layer1.output.grad.save()

        loss = model.output.sum()
        loss.backward()

    print("L1_Grad_Before: ", l1_grad_before)
    print("L1_Grad_Before_Address: ", hex(id(l1_grad_before)))
    print("L1_Grad_After: ", l1_grad_after)
    print("L1_Grad_After_Address: ", hex(id(l1_grad_after)))
```

Out:
```
L1_Grad_Before:  tensor([[-0.0336, -0.0316,  0.5292, -0.1040, -0.0569,  0.0234, -0.1658, -0.3092,
          0.0051,  0.3399]])
L1_Grad_Before_Address:  0x11f3c1ea0
L1_Grad_After:  tensor([[-0.0336, -0.0316,  0.5292, -0.1040, -0.0569,  0.0234, -0.1658, -0.3092,
          0.0051,  0.3399]])
L1_Grad_After_Address:  0x11f3c1ea0
```

The **second example** shows that modifying directly the `.data` attribute of the Grad Proxy does not bring up the bug and produces the correct expected results. This occurs because setting `.data` does not involve a `"swap"` of the Grad Proxy since `.data` is not a Proxy itself but just an attribute. This is also shown by the intervention graph.

Code:
``` Python3
with model.trace(input):
        model.layer1.output.requires_grad = True

        l1_grad_before = model.layer1.output.grad.clone().save()

        model.layer1.output.grad.data = model.layer1.output.grad.clone() * 2

        l1_grad_after = model.layer1.output.grad.save()

        loss = model.output.sum()
        loss.backward()

    print("L1_Grad_Before: ", l1_grad_before)
    print("L1_Grad_After: ", l1_grad_after)
```

Out:
```
L1_Grad_Before:  tensor([[-0.0336, -0.0316,  0.5292, -0.1040, -0.0569,  0.0234, -0.1658, -0.3092,
          0.0051,  0.3399]])
L1_Grad_After:  tensor([[-0.0673, -0.0633,  1.0584, -0.2081, -0.1137,  0.0469, -0.3317, -0.6185,
          0.0102,  0.6798]])
```

Intervention Graph:
![Layer 1 Grad Setting Data Attr](https://github.com/ndif-team/nnsight/assets/57205266/db38238a-8be3-4cdf-a2dc-b99494979123)

**Comment by JadenFiotto-Kaufman** [comment]
@AdamBelfki3  Is this fixed?

**Comment by AdamBelfki3** [comment]
Yes it is!
